src/v/security/jwt.h

```diff
@@ -0,0 +1,623 @@
+      : _impl(std::move(v)) {}
+
+    auto alg() {
+        return ss::visit(_impl, [](auto const& impl) { return impl.alg(); });
+    }
+
+    auto kty() {
+        return ss::visit(_impl, [](auto const& impl) { return impl.kty(); });
+    }
+
+    bool verify(cryptopp_byte_view msg, cryptopp_byte_view sig) {
+        return ss::visit(
+          _impl, [=](auto const& impl) { return impl.verify(msg, sig); });
+    }
+
+private:
+    using verifier_impls = std::variant<rs256_verifier>;
+    verifier_impls _impl;
+};
+
+inline result<verifier> make_rs256_verifier(
+  json::Value const& jwk, CryptoPP::AutoSeededRandomPool& rng) {
+    CryptoPP::RSA::PublicKey key;
+    try {
+        auto n = detail::as_cryptopp_integer(jwk, "n");
+        auto e = detail::as_cryptopp_integer(jwk, "e");
+        if (!n.has_value() || !e.has_value()) {
+            return errc::jwk_invalid;
+        }
+        key.Initialize(n.value(), e.value());
+        if (!key.Validate(rng, 3)) {
+            return errc::jwk_invalid;
+        }
+    } catch (CryptoPP::Exception const& ex) {
+        return errc::jwk_invalid;
+    }
+    return verifier{rs256_verifier{key}};
+}
+
+using verifiers = absl::flat_hash_map<
+  ss::sstring,
+  detail::verifier,
+  detail::string_viewable_hasher,
+  detail::string_viewable_compare>;
+
+inline result<verifiers>
+make_verifiers(jwks const& jwks, CryptoPP::AutoSeededRandomPool& rng) {
+    auto keys = jwks.keys();
+    verifiers vs;
+    for (auto const& key : keys) {
+        // Alg is required
+        auto alg = detail::string_view(key, "alg");
+        if (!alg) {
+            return errc::jwk_invalid;
+        }
+
+        // Use is optional, but must be sig if it exists
+        auto use = detail::string_view(key, "use");
+        if (use.value_or("sig") != "sig") {
+            continue;
+        }
+
+        using factory = result<verifier> (*)(
+          json::Value const&, CryptoPP::AutoSeededRandomPool&);
+        auto v = string_switch<std::optional<factory>>(*alg)
+                   .match(rs256_str, &make_rs256_verifier)
+                   .default_match(std::optional<factory>{});
+        if (!v) {
+            continue;
+        }
+
+        auto r = v.value()(key, rng);
+        if (!r) {
+            continue;
+        }
+
+        vs.emplace(
+          detail::string_view(key, "kid").value_or(""),
+          std::move(r).assume_value());
+    }
+    if (vs.empty()) {
+        return errc::jwks_invalid;
+    }
+    return vs;
+}
+
+} // namespace detail
+
+class oidc_verifier {
+public:
+    explicit oidc_verifier() = default;
+
+    // Verify the JWS signature and return the JWT
+    result<jwt> verify(jws const& jws) {
+        std::string_view sv(jws._encoded);
+
+        if (!detail::check_bytes(sv)) {
+            return errc::jws_invalid;
+        }
+
+        std::vector<detail::cryptopp_byte_view> jose_enc;
+        jose_enc.reserve(3);
+        boost::algorithm::split(
```

Comment:
Could also use absl::StrSplit which would be a bit simpler: https://abseil.io/docs/cpp/guides/strings#abslstrsplit-for-splitting-strings

```
std::vector<std::string_view> parts = absl::StrSplit(sv, '.');
```

Reply:
> Could also use absl::StrSplit which would be a bit simpler: https://abseil.io/docs/cpp/guides/strings#abslstrsplit-for-splitting-strings
> 
> ```
> std::vector<std::string_view> parts = absl::StrSplit(sv, '.');
> ```

It's a nicer interface, but can't deal with `CryptoPP::byte` (`unsigned char`). I can split in `char` and cast the parts at the call to `base64_url_decode`.

The casting everywhere is a little frustrating.

Reply:
Bummer.